### PR TITLE
ESQL: Remove SearchContext from Operators

### PR DESCRIPTION
**Pull request by nik9000**
This removes the *giant* `SearchContext` class from our `Operator`s. It's just *so* **so** ***so*** big. It's hard to test with. You can do it. There's a lot of crazy mocking. Lots of reading.

The replacement is two interfaces, on in the compute engine, called `ShardContext`. It contains the methods that are required to power all of the `LuceneSourceOperator`s. It has the `IndexSearcher`, can build sorts, and has a few identifiers.

The other interface is also called `ShardContext`, but this one lives in esql itself and it has all of the things needed to power the `EsPhysicalOperationProviders` - mostly stuff to configure the value fetching operator and build queries.

There exists a "production" implementation of these interfaces in `EsPhysicalOperationProviders` and a test implementation in `LuceneSourceOperatorTests`. It's super easy to plug in non-production implementations that don't drag in 45% of all of Elasticsearch.

### CODE REVIEW

**Comment by nik9000** [comment]
It's certainly possible to keep using `SearchContext` everywhere. But it requires so much reading!

**Review by costin — APPROVED** [review]
Great stuff! Abstracting the low-level APIs significantly improves the code and improves testability (e.g. SearchStats so we could run tests without mocking Lucene).

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneSourceOperator.java [hunk 1]
```diff
@@ -149,7 +148,7 @@ public Page getOutput() {
                 IntBlock leaf = null;
                 IntVector docs = null;
                 try {
-                    shard = blockFactory.newConstantIntBlockWith(scorer.shardIndex(), currentPagePos);
+                    shard = blockFactory.newConstantIntBlockWith(scorer.shardContext().index(), currentPagePos);
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneSourceOperator.java:151]: thought: I was confused why using `.shardId()` is incorrect here, but the tests for `metadata _id` don't pass with this.

**Review by alex-spies — APPROVED** [review]
This is fantastic, thanks a lot Nik!

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneSourceOperator.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneSourceOperator.java:151]: Yeah - `shardId` is the count of the shard in the list of shards for the current index. Like, if you do `_cat/shards` it's the number that shows next to the shard. So, like, "this is the 0th shard of the index named foo". `index` is the index into the list. I'll see if i can make this more readable.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ShardContext.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.Optional;
+
+/**
+ * Context of each shard we're operating against.
+ */
+public interface ShardContext {
+    /**
+     * The index of this shard in the list of shards being processed.
+     */
+    int index();
```
[the remark this answers is not in the document]
Comment (alex-spies) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ShardContext.java:25]: It's great now as it is, thanks for the iteration!

**Comment by nik9000** [comment]
All merged up! Good luck building on it @alex-spies .

**Pull request merged**